// @ionic-native/streaming-media/__ivy_ngcc__/ngx/index.js
import { __decorate, __extends } from "tslib";
import { Injectable } from '@angular/core';
import { IonicNativePlugin, cordova } from '@ionic-native/core';
import * as ɵngcc0 from '@angular/core';
var StreamingMedia = /** @class */ (function (_super) {
    __extends(StreamingMedia, _super);
    function StreamingMedia() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    StreamingMedia.prototype.playVideo = function (videoUrl, options) { return cordova(this, "playVideo", { "sync": true }, arguments); };
    StreamingMedia.prototype.playAudio = function (audioUrl, options) { return cordova(this, "playAudio", { "sync": true }, arguments); };
    StreamingMedia.prototype.stopAudio = function () { return cordova(this, "stopAudio", { "sync": true }, arguments); };
    StreamingMedia.prototype.pauseAudio = function () { return cordova(this, "pauseAudio", { "sync": true, "platforms": ["iOS"] }, arguments); };
    StreamingMedia.prototype.resumeAudio = function () { return cordova(this, "resumeAudio", { "sync": true, "platforms": ["iOS"] }, arguments); };
    StreamingMedia.pluginName = "StreamingMedia";
    StreamingMedia.plugin = "cordova-plugin-streaming-media";
    StreamingMedia.pluginRef = "plugins.streamingMedia";
    StreamingMedia.repo = "https://github.com/nchutchind/cordova-plugin-streaming-media";
    StreamingMedia.platforms = ["Amazon Fire OS", "Android", "iOS"];
StreamingMedia.ɵfac = function StreamingMedia_Factory(t) { return ɵStreamingMedia_BaseFactory(t || StreamingMedia); };
StreamingMedia.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: StreamingMedia, factory: function (t) { return StreamingMedia.ɵfac(t); } });
var ɵStreamingMedia_BaseFactory = /*@__PURE__*/ ɵngcc0.ɵɵgetInheritedFactory(StreamingMedia);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(StreamingMedia, [{
        type: Injectable
    }], null, null); })();
    return StreamingMedia;
}(IonicNativePlugin));
export { StreamingMedia };
